File: src/types/game.ts
/**
 * Frontend mirror of the Rust domain types. Keep these field names in sync with
 * `src-tauri/src/domain/game.rs` (serde uses snake_case on the wire).
 */

export const GAME_STATUSES = [
  "planned",
  "playing",
  "paused",
  "completed",
  "completed_100",
  "dropped",
] as const;

export type GameStatus = (typeof GAME_STATUSES)[number];

export const STATUS_LABELS: Record<GameStatus, string> = {
  planned: "Planned",
  playing: "Playing",
  paused: "Paused",
  completed: "Completed",
  completed_100: "Completed 100%",
  dropped: "Dropped",
};

export interface Game {
  id: number;
  title: string;
  genres: string[];
  platform: string | null;
  developer: string | null;
  publisher: string | null;
  releaseYear: number | null;
  startedAt: string | null;
  finishedAt: string | null;
  status: GameStatus;
  rating: number | null;
  coverPath: string | null;
  createdAt: string;
  updatedAt: string;
}

/** Payload for creating or updating a game (everything except server-managed fields). */
export interface GameInput {
  title: string;
  genres: string[];
  platform: string | null;
  developer: string | null;
  publisher: string | null;
  releaseYear: number | null;
  startedAt: string | null;
  finishedAt: string | null;
  status: GameStatus;
  rating: number | null;
  coverPath: string | null;
}

export type GenreMatch = "any" | "all";

export const GAME_SORTS = [
  "default",
  "title",
  "started_at",
  "finished_at",
  "release_year",
  "rating",
  "status",
] as const;
export type GameSort = (typeof GAME_SORTS)[number];

export const SORT_LABELS: Record<GameSort, string> = {
  default: "Default (Playing first)",
  title: "Title",
  started_at: "Started",
  finished_at: "Finished",
  release_year: "Release year",
  rating: "Rating",
  status: "Status",
};

export type SortDirection = "asc" | "desc";

/**
 * Flexible game browser query. Mirrors the Rust `GameQuery`. Only set the fields
 * you want to filter on; omit the rest. Combine filters freely.
 */
export interface GameQuery {
  search?: string;
  statuses?: GameStatus[];
  platforms?: string[];
  genres?: string[];
  genreMatch?: GenreMatch;
  releaseYear?: number;
  startedYear?: number;
  finishedYear?: number;
  playedYear?: number;
  minRating?: number;
  maxRating?: number;
  sort?: GameSort;
  direction?: SortDirection;
}

File: src-tauri/src/domain/game.rs
use serde::{Deserialize, Serialize};

/// Where a game sits in the player's journey. The wire/storage representation is
/// the snake_case string (e.g. `completed_100`).
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum GameStatus {
    Planned,
    Playing,
    Paused,
    Completed,
    #[serde(rename = "completed_100")]
    Completed100,
    Dropped,
}

impl GameStatus {
    /// Canonical string used both on the wire and in the database.
    pub fn as_str(self) -> &'static str {
        match self {
            GameStatus::Planned => "planned",
            GameStatus::Playing => "playing",
            GameStatus::Paused => "paused",
            GameStatus::Completed => "completed",
            GameStatus::Completed100 => "completed_100",
            GameStatus::Dropped => "dropped",
        }
    }

    pub fn parse(value: &str) -> Option<Self> {
        Some(match value {
            "planned" => GameStatus::Planned,
            "playing" => GameStatus::Playing,
            "paused" => GameStatus::Paused,
            "completed" => GameStatus::Completed,
            "completed_100" => GameStatus::Completed100,
            "dropped" => GameStatus::Dropped,
            _ => return None,
        })
    }
}

/// A game as stored and returned to the UI. Field names cross the boundary in
/// camelCase to match the TypeScript `Game` interface.
#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct Game {
    pub id: i64,
    pub title: String,
    pub genre: Option<String>,
    pub platform: Option<String>,
    pub developer: Option<String>,
    pub publisher: Option<String>,
    pub release_year: Option<i64>,
    pub started_at: Option<String>,
    pub finished_at: Option<String>,
    pub status: GameStatus,
    pub rating: Option<i64>,
    pub cover_path: Option<String>,
    pub created_at: String,
    pub updated_at: String,
}

/// User-supplied fields for creating or updating a game. Server-managed columns
/// (`id`, timestamps) are intentionally absent.
#[derive(Debug, Clone, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct GameInput {
    pub title: String,
    pub genre: Option<String>,
    pub platform: Option<String>,
    pub developer: Option<String>,
    pub publisher: Option<String>,
    pub release_year: Option<i64>,
    pub started_at: Option<String>,
    pub finished_at: Option<String>,
    pub status: GameStatus,
    pub rating: Option<i64>,
    pub cover_path: Option<String>,
}
